python_modules/dagster/dagster/_core/definitions/new_freshness_policy.py

```diff
@@ -0,0 +1,93 @@
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Optional
+
+from dagster import _check as check
+from dagster._annotations import beta
+from dagster._record import IHaveNew, record_custom
+from dagster._serdes import whitelist_for_serdes
+
+
+class FreshnessPolicyType(Enum):
+    TIME_WINDOW = "time_window"
+    CRON = "cron"
+
+
+class NewFreshnessPolicy(ABC):
+    """A condition that delineates when an asset is considered fresh. The asset can either pass, violate, or warn on this condition.
+
+    This is intended to replace the existing, deprecated `FreshnessPolicy` construct.
+
+    `NewFreshnessPolicy` can be specified as a parameter to `@asset`, `@multi_asset`, or in `AssetSpec`.
+
+    Do not use this class directly. Instead, use one of the subclasses.
+    """
+
+    @property
+    @abstractmethod
+    def policy_type(self) -> FreshnessPolicyType: ...
+
+
+@beta
+@whitelist_for_serdes
+@record_custom
+class TimeWindowFreshnessPolicy(NewFreshnessPolicy, IHaveNew):
+    """[BETA/EXPERIMENTAL] A freshness condition that considers an asset fresh if it was materialized within a specified time window (time_window_minutes).
+
+    The asset is considered stale if its last materialization is older than time_window_minutes, and in a warning state
+    if its last materialization is older than the warning time window but newer than the critical time window.
+
+    Example usage:
+
+    .. code-block:: python
+
+        from dagster import TimeWindowFreshnessPolicy
+
+        # Asset must be materialized within last 24 hours (1440 minutes)
+        @asset(new_freshness_policy=TimeWindowFreshnessPolicy(time_window_minutes=1440))
+        def my_asset():
+            ...
+
+        # Asset must be materialized within last 24 hours (1440 minutes), with warning after 12 hours
+        @asset(new_freshness_policy=TimeWindowFreshnessPolicy(time_window_minutes=1440, warning_time_window_minutes=720))
+        def my_asset():
+            ...
+
+    Args:
+        time_window_minutes (int): The maximum age in minutes for an asset to be considered fresh.
+            Must be positive.
+        warning_time_window_minutes (Optional[int]): If provided, the age in minutes at which to start
+            showing warnings about asset freshness. Must be positive and less than time_window_minutes.
+    """
+
+    time_window_minutes: int
+    warning_time_window_minutes: Optional[int] = None
+
+    @property
+    def policy_type(self) -> FreshnessPolicyType:
+        return FreshnessPolicyType.TIME_WINDOW
```

Comment:
this can happen in a downstream PR (probably better it does actually, as this already has a decent bit of content), but the actual "freshness evaluation method" should definitely be part of this class. In addition to the version skew considerations owen mentioned, you can imagine wanting to allow users to call this directly from user code for example in order to define their own user-code based freshness, as in the product review.